C2Rust generates invalid Rust by taking address of a temporary when translating wide string literals (E0745)

C2Rust generates **invalid Rust code** when translating C wide string literals (L"...").
The generated code takes the address of a **temporary value** produced by `core::mem::transmute`, which fails to compile with:

```
error[E0745]: cannot take address of a temporary
```

In C, wide string literals have **static storage duration**, but the generated Rust code does not preserve this property.
**Original C Program**

```c
#include <stdio.h>
#include <wchar.h>

int main() {
    const wchar_t* wstr = L"Wide String";
    printf("Wide string, First char code=%d\n", (int)wstr[0]);
    return 0;
}
```

**C2Rust-Generated Rust**

```rust
#![allow(
    dead_code,
    non_camel_case_types,
    non_snake_case,
    non_upper_case_globals,
    unused_assignments,
    unused_mut
)]
#![feature(raw_ref_op)]
extern "C" {
    fn printf(__format: *const ::core::ffi::c_char, ...) -> ::core::ffi::c_int;
}
pub type wchar_t = ::libc::wchar_t;
unsafe fn main_0() -> ::core::ffi::c_int {
    let mut wstr: *const wchar_t = &raw const ::core::mem::transmute::<
        [u8; 48],
        [::core::ffi::c_int; 12],
    >(
        *b"W\0\0\0i\0\0\0d\0\0\0e\0\0\0 \0\0\0S\0\0\0t\0\0\0r\0\0\0i\0\0\0n\0\0\0g\0\0\0\0\0\0\0",
    ) as *const wchar_t;
    printf(
        b"Wide string, First char code=%d\n\0" as *const u8
            as *const ::core::ffi::c_char,
        *wstr.offset(0 as ::core::ffi::c_int as isize) as ::core::ffi::c_int,
    );
    return 0 as ::core::ffi::c_int;
}
pub fn main() {
    unsafe { ::std::process::exit(main_0() as i32) }
}
```

**Observed Behavior**
Compilation fails with:

```
error[E0745]: cannot take address of a temporary
```

because `&raw const transmute(...)` takes the address of a **temporary value**, whose lifetime ends immediately.
**Environment**
- c2rust version: v0.21.0
- rustc version: 1.91.1

--- a/c2rust-transpile/src/translator/literals.rs
+++ b/c2rust-transpile/src/translator/literals.rs
@@ -119,16 +119,29 @@ impl<'c> Translation<'c> {
 
                 if ctx.needs_address && element_size == 1 {
                     // Unlike in C, Rust string literals are already references by default.
-                    // So if the address needs to be taken, just make a bare literal.
+                    // So if the address needs to be taken, just make a bare literal and let
+                    // `convert_address_of_common` cast it to the appropriate type.
+                    // Strings with element_size > 1 cannot be cast from a byte literal for
+                    // alignment reasons, and need a transmute.
                     Ok(WithStmts::new_val(val))
                 } else {
                     // std::mem::transmute::<[u8; size], ctype>(*b"xxxx")
                     let array_ty = mk().array_ty(mk().ident_ty("u8"), mk().lit_expr(len as u128));
-                    let val = transmute_expr(
+                    let mut val = transmute_expr(
                         array_ty,
                         self.convert_type(ty.ctype)?,
                         mk().unary_expr(UnOp::Deref(Default::default()), val),
                     );
+
+                    // A transmute creates a temporary, which cannot have its address taken without
+                    // creating dangling pointers. Wrap it inside an inline `const` block, so that
+                    // it will be const-promoted to 'static.
+                    if ctx.needs_address {
+                        self.use_feature("inline_const");
+                        let stmts = vec![mk().expr_stmt(val)];
+                        val = mk().const_block_expr(mk().const_block(stmts));
+                    }
+
                     Ok(WithStmts::new_unsafe_val(val))
                 }
             }

--- a/c2rust-transpile/src/translator/mod.rs
+++ b/c2rust-transpile/src/translator/mod.rs
@@ -1580,8 +1580,10 @@ impl<'c> Translation<'c> {
 
     /// Called when translation makes use of a language feature that will require a feature-gate.
     pub fn use_feature(&self, feature: &'static str) {
-        if matches!(feature, "asm" | "label_break_value" | "raw_ref_op")
-            && self.tcfg.edition >= Edition2024
+        if matches!(
+            feature,
+            "asm" | "inline_const" | "label_break_value" | "raw_ref_op"
+        ) && self.tcfg.edition >= Edition2024
         {
             return;
         }

--- a/c2rust-transpile/src/translator/pointers.rs
+++ b/c2rust-transpile/src/translator/pointers.rs
@@ -138,15 +138,11 @@ impl<'c> Translation<'c> {
             }
             needs_cast = true;
         }
-        // Values that translate into temporaries can't be raw-borrowed in Rust,
-        // and must be regular-borrowed first.
-        // Borrowing in a const context will extend the lifetime to static.
+        // Values that translate into const temporaries can't be raw-borrowed in Rust.
+        // They must be regular-borrowed first, which will extend the lifetime to static.
         else if arg_is_macro
-            || ctx.is_const
-                && matches!(
-                    arg_expr_kind,
-                    Some(CExprKind::Literal(..) | CExprKind::CompoundLiteral(..))
-                )
+            || matches!(arg_expr_kind, Some(CExprKind::Literal(..)))
+            || ctx.is_const && matches!(arg_expr_kind, Some(CExprKind::CompoundLiteral(..)))
         {
             let arg_cty_kind = &self.ast_context.resolve_type(arg_cty.ctype).kind;
 

--- a/c2rust-transpile/tests/snapshots/os-specific/wide_strings.c
+++ b/c2rust-transpile/tests/snapshots/os-specific/wide_strings.c
@@ -3,16 +3,13 @@
 wchar_t static_array[] = L"x";
 wchar_t static_array_longer[3] = L"x";
 wchar_t static_array_shorter[1] = L"xy";
-// Currently broken, see #1571
-// const wchar_t *static_ptr = L"x";
+const wchar_t *static_ptr = L"x";
 
 void func() {
     wchar_t array[] = L"x";
     wchar_t array_longer[3] = L"x";
     wchar_t array_shorter[1] = L"xy";
-
-    // Currently broken, see #1571
-    // const wchar_t *ptr = L"x";
+    const wchar_t *ptr = L"x";
 
     size_t len = wcslen(array);
 }

